lib/kitchen/driver/gce.rb

```diff
@@ -16,154 +17,507 @@
       end
 
-      private
+      def validate!
+        raise "Either zone or region must be specified" unless config[:zone] || config[:region]
+        raise "'any' is no longer a valid region" if config[:region] == "any"
+        raise "Zone #{config[:zone]} is not a valid zone" if config[:zone] && !valid_zone?
+        raise "Region #{config[:region]} is not a valid region" if config[:region] && !valid_region?
+        raise "Machine type #{config[:machine_type]} is not valid" unless valid_machine_type?
+        raise "Disk type #{config[:disk_type]} is not valid" unless valid_disk_type?
+        raise "Network #{config[:network]} is not valid" unless valid_network?
+        raise "Email address of GCE user is not set" if winrm_transport? && config[:email].nil?
+
+        warn("Both zone and region specified - region will be ignored.") if config[:zone] && config[:region]
+      end
 
       def connection
-        options = {
-          provider: 'google',
-          google_client_email: config[:google_client_email],
-          google_project: config[:google_project]
-        }
+        return @connection unless @connection.nil?
 
-        [
-          :google_key_location,
-          :google_json_key_location
-        ].each do |k|
-          options[k] = config[k] unless config[k].nil?
+        @connection = Google::Apis::ComputeV1::ComputeService.new
+        @connection.authorization = authorization
+        @connection.client_options = Google::Apis::ClientOptions.new.tap do |opts|
+          opts.application_name    = "kitchen-google"
+          opts.application_version = Kitchen::Driver::GCE_VERSION
         end
 
-        Fog::Compute.new(options)
+        @connection
       end
 
-      def create_disk
-        disk = connection.disks.create(
-          name: config[:inst_name],
-          size_gb: config[:disk_size],
-          zone_name: config[:zone_name],
-          source_image: config[:image_name]
+      def authorization
+        @authorization ||= Google::Auth.get_application_default(
+          [
+            "https://www.googleapis.com/auth/cloud-platform",
+            "https://www.googleapis.com/auth/compute",
+          ]
         )
+      end
+
+      def winrm_transport?
+        instance.transport.name.downcase == "winrm"
+      end
+
+      def update_windows_password(server_name)
+        return unless winrm_transport?
+
+        username = instance.transport[:username]
+
+        info("Resetting the Windows password for user #{username} on #{server_name}...")
+
+        state[:password] = GoogleComputeWindowsPassword.new(
+          project:       project,
+          zone:          zone,
+          instance_name: server_name,
+          email:         config[:email],
+          username:      username
+        ).new_password
+
+        info("Password reset complete on #{server_name} complete.")
+      end
+
+      def check_api_call(&block)
+        block.call
+      rescue Google::Apis::ClientError
```

Comment:
IMO we should at least return the human-readable part of the call as error here.

Otherwise, what happens if we don't correctly configure notification:
- User incorrectly configures the project (esp. since `project` parameter is a bit ambiguous).
- Kitchen falls through to this `check_api_call` method.
- Block gets called:  `check_api_call { connection.get_zone(project, config[:zone]) }`
- `connection.get_zone` responds with:
```
Google::Apis::ClientError: notFound: The resource 'projects/xxx' was not found
```
- Code falls through with `false` and the *end user* gets:
```
-----> Starting Kitchen (v1.6.0)
-----> Creating <default-debian-7>...
>>>>>> Error encountered during server creation: RuntimeError: Zone europe-west1-d is not a valid zone
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: Failed to complete #create action: [Zone europe-west1-d is not a valid zone]
```
, which is completely off the mark in my opinion.

What do you think?